File: src/CircleAI.Inference/PrefixCacheService.cs
// PrefixCacheService.cs
//
// RT-06: cross-session prefix cache. The motivation:
//
// Every fresh conversation today re-pays the system-prompt prefill cost — on a
// Tier-0 device that's 2-3 seconds before the first token appears. The user
// perceives "slow." But the system prompt is almost always identical across
// chats with the same persona / app. So: snapshot the model's KV state once
// per (modelId, systemPrompt) pair, reload it on the next chat with the same
// pair, and skip the prefill entirely.
//
// The implementation uses MNN's existing mnn_llm_save_session / load_session
// primitives — no native bridge changes needed. The on-disk format is whatever
// MNN itself writes; we only own the indexing.
//
// Cache layout:
//   %LOCALAPPDATA%/CircleAI/prefix-cache/
//     <modelHash>_<systemHash>.session   ← MNN-native KV snapshot
//     <modelHash>_<systemHash>.meta      ← JSON metadata (createdAtUtc, modelId)
//
// Eviction policy v1: simple LRU by file mtime, cap at 500 MB total, evict the
// oldest first. Tier-0 phones get hit hard; the cap matters.

using System;
using System.IO;
using System.Linq;
using System.Security.Cryptography;
using System.Text;
using System.Threading;
using System.Threading.Tasks;

namespace CircleAI.Inference;

/// <summary>
/// Manages an on-disk cache of "warm" model sessions keyed by the hash of
/// (modelId, systemPrompt). Generators that opt in via
/// <see cref="GenerationOptions.UsePrefixCache"/> consult this service before
/// resetting the model handle for a new conversation.
/// <para>
/// The service is thread-safe and shared across generators; default instance
/// is <see cref="Default"/>. Override the root directory via the constructor
/// for tests.
/// </para>
/// </summary>
public sealed class PrefixCacheService
{
    private const int CapBytes = 500 * 1024 * 1024; // 500 MB
    private static readonly SemaphoreSlim _ioLock = new(1, 1);

    private readonly string _root;

    /// <summary>
    /// The default per-app instance rooted at
    /// <c>%LOCALAPPDATA%/CircleAI/prefix-cache</c> on Windows and
    /// <c>~/.circleai/prefix-cache</c> on Unix / iOS / Android (the platform's
    /// home directory equivalent).
    /// </summary>
    public static PrefixCacheService Default { get; } = new(DefaultRoot());

    /// <summary>
    /// Construct a cache service rooted at <paramref name="root"/>. The directory
    /// is created on demand.
    /// </summary>
    public PrefixCacheService(string root)
    {
        if (string.IsNullOrWhiteSpace(root))
            throw new ArgumentException("root is required.", nameof(root));
        _root = root;
        Directory.CreateDirectory(_root);
    }

    /// <summary>
    /// Compute the cache key for a (modelId, systemPrompt) pair. Returns
    /// <c>null</c> when <paramref name="systemPrompt"/> is null/empty — there
    /// is nothing to cache without a system prompt to key against.
    /// </summary>
    public static string? KeyFor(string modelId, string? systemPrompt)
    {
        if (string.IsNullOrWhiteSpace(modelId))    return null;
        if (string.IsNullOrEmpty(systemPrompt))    return null;

        var modelHash  = Sha256(modelId);
        var systemHash = Sha256(systemPrompt!);
        // First 16 hex chars per component — collision-free at the scale of
        // any single device's cache (≪ 10⁶ entries), much shorter on disk.
        return $"{modelHash[..16]}_{systemHash[..16]}";
    }

    /// <summary>
    /// Returns the cache path for <paramref name="key"/>. The returned path
    /// may or may not exist; use <see cref="HasEntryAsync"/> to check.
    /// </summary>
    public string PathFor(string key) => Path.Combine(_root, $"{key}.session");

    /// <summary>
    /// <c>true</c> when a cached entry exists for <paramref name="key"/>.
    /// </summary>
    public Task<bool> HasEntryAsync(string key, CancellationToken ct = default)
        => Task.FromResult(File.Exists(PathFor(key)));

    /// <summary>
    /// Touch the entry's mtime so LRU eviction treats it as recently used.
    /// Called after a successful load.
    /// </summary>
    public void Touch(string key)
    {
        var path = PathFor(key);
        if (File.Exists(path))
            File.SetLastWriteTimeUtc(path, DateTime.UtcNow);
    }

    /// <summary>
    /// Evict oldest entries until the directory is under
    /// <see cref="CapBytes"/>. Called after every successful Save to keep the
    /// cache bounded. Best-effort — failures are swallowed.
    /// </summary>
    public async Task EvictIfNeededAsync(CancellationToken ct = default)
    {
        await _ioLock.WaitAsync(ct).ConfigureAwait(false);
        try
        {
            var dir = new DirectoryInfo(_root);
            if (!dir.Exists) return;

            var files = dir.EnumerateFiles("*.session")
                           .OrderBy(f => f.LastWriteTimeUtc)
                           .ToList();

            long total = files.Sum(f => f.Length);
            int i = 0;
            while (total > CapBytes && i < files.Count)
            {
                var f = files[i++];
                try { total -= f.Length; f.Delete(); }
                catch { /* best effort */ }
            }
        }
        finally { _ioLock.Release(); }
    }

    // ── helpers ──────────────────────────────────────────────────────────

    private static string Sha256(string input)
    {
        var bytes = SHA256.HashData(Encoding.UTF8.GetBytes(input));
        var sb    = new StringBuilder(bytes.Length * 2);
        foreach (var b in bytes) sb.Append(b.ToString("x2"));
        return sb.ToString();
    }

    private static string DefaultRoot()
    {
        // Windows: %LOCALAPPDATA%/CircleAI/prefix-cache
        // Unix-like (Linux/macOS/Android/iOS): ~/.circleai/prefix-cache
        var local = Environment.GetEnvironmentVariable("LOCALAPPDATA");
        if (!string.IsNullOrWhiteSpace(local))
            return Path.Combine(local!, "CircleAI", "prefix-cache");

        var home = Environment.GetFolderPath(Environment.SpecialFolder.UserProfile);
        return Path.Combine(home, ".circleai", "prefix-cache");
    }
}

File: kotlin/src/main/kotlin/com/bhengubv/circleai/inference/PrefixCacheService.kt
// PrefixCacheService.kt
//
// Kotlin port of CircleAI.Inference.PrefixCacheService (RT-06 cross-session
// prefix cache). C# is the EXACT spec. Manages an on-disk cache of "warm"
// model sessions keyed by the hash of (modelId, systemPrompt). Generators that
// opt in via GenerationOptions.usePrefixCache consult this service before
// resetting the model handle for a new conversation.
//
// Cache layout:
//   %LOCALAPPDATA%/CircleAI/prefix-cache/    (Windows)
//   ~/.circleai/prefix-cache/                (Unix/iOS/Android)
//     <modelHash>_<systemHash>.session   ← native KV snapshot
//     <modelHash>_<systemHash>.meta      ← JSON metadata
//
// Eviction: LRU by file mtime, cap at 500 MB total, oldest first.

package com.bhengubv.circleai.inference

import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock
import java.io.File
import java.security.MessageDigest
import java.time.Instant

/**
 * Manages an on-disk cache of "warm" model sessions keyed by the hash of
 * (modelId, systemPrompt).
 *
 * The service is thread-safe (I/O serialised through a [Mutex]) and shared
 * across generators; the default instance is [Default]. Override the root
 * directory via the constructor for tests.
 *
 * @param root Directory to root the cache at. Created on demand.
 */
class PrefixCacheService(root: String) {

    private val root: String

    init {
        require(root.isNotBlank()) { "root is required." }
        this.root = root
        File(root).mkdirs()
    }

    /** Returns the cache path for [key]. May or may not exist; use [hasEntryAsync]. */
    fun pathFor(key: String): String = File(root, "$key.session").path

    /** `true` when a cached entry exists for [key]. */
    suspend fun hasEntryAsync(key: String): Boolean = File(pathFor(key)).exists()

    /**
     * Touch the entry's mtime so LRU eviction treats it as recently used.
     * Called after a successful load.
     */
    fun touch(key: String) {
        val f = File(pathFor(key))
        if (f.exists()) f.setLastModified(System.currentTimeMillis())
    }

    /**
     * Evict oldest entries until the directory is under [CAP_BYTES]. Called
     * after every successful save to keep the cache bounded. Best-effort —
     * failures are swallowed.
     */
    suspend fun evictIfNeededAsync() {
        ioLock.withLock {
            val dir = File(root)
            if (!dir.exists()) return
            val files = (dir.listFiles { f -> f.isFile && f.name.endsWith(".session") } ?: emptyArray())
                .sortedBy { it.lastModified() }
            var total = files.sumOf { it.length() }
            var i = 0
            while (total > CAP_BYTES && i < files.size) {
                val f = files[i++]
                try {
                    total -= f.length()
                    f.delete()
                } catch (_: Exception) {
                    // best effort
                }
            }
        }
    }

    companion object {
        private const val CAP_BYTES: Long = 500L * 1024 * 1024 // 500 MB
        private val ioLock = Mutex()

        /**
         * The default per-app instance rooted at `%LOCALAPPDATA%/CircleAI/prefix-cache`
         * on Windows and `~/.circleai/prefix-cache` on Unix.
         */
        val Default: PrefixCacheService by lazy { PrefixCacheService(defaultRoot()) }

        /**
         * Compute the cache key for a (modelId, systemPrompt) pair. Returns
         * `null` when [systemPrompt] is null/empty — there is nothing to cache
         * without a system prompt to key against.
         */
        fun keyFor(modelId: String, systemPrompt: String?): String? {
            if (modelId.isBlank()) return null
            if (systemPrompt.isNullOrEmpty()) return null
            val modelHash = sha256(modelId)
            val systemHash = sha256(systemPrompt)
            // First 16 hex chars per component — collision-free at device scale.
            return "${modelHash.substring(0, 16)}_${systemHash.substring(0, 16)}"
        }

        private fun sha256(input: String): String {
            val bytes = MessageDigest.getInstance("SHA-256").digest(input.toByteArray(Charsets.UTF_8))
            val sb = StringBuilder(bytes.size * 2)
            for (b in bytes) sb.append("%02x".format(b))
            return sb.toString()
        }

        private fun defaultRoot(): String {
            val local = System.getenv("LOCALAPPDATA")
            if (!local.isNullOrBlank()) {
                return File(File(local, "CircleAI"), "prefix-cache").path
            }
            val home = System.getProperty("user.home") ?: "."
            return File(File(home, ".circleai"), "prefix-cache").path
        }
    }
}

/**
 * Metadata sidecar for a cached prefix entry (matches the C# `.meta` JSON:
 * createdAtUtc + modelId). Written alongside the `.session` snapshot.
 */
data class PrefixCacheMeta(
    val modelId: String,
    val createdAtUtc: Instant,
)
